libs/native/src/main21/java/org/elasticsearch/nativeaccess/jdk/JdkVectorLibrary.java

```diff
@@ -23,142 +24,166 @@
 
 public final class JdkVectorLibrary implements VectorLibrary {
 
+    static final VectorSimilarityFunctions INSTANCE;
+
     static {
         System.loadLibrary("vec");
+        final MethodHandle vecCaps$mh = downcallHandle("vec_caps", FunctionDescriptor.of(JAVA_INT));
+
+        try {
+            int caps = (int) vecCaps$mh.invokeExact();
+            if (caps != 0) {
+                INSTANCE = new JdkVectorSimilarityFunctions();
+            } else {
+                INSTANCE = null;
+            }
+        } catch (Throwable t) {
+            throw new AssertionError(t);
+        }
     }
 
     public JdkVectorLibrary() {}
 
-    static final MethodHandle dot8stride$mh = downcallHandle("dot8s_stride", FunctionDescriptor.of(JAVA_INT));
-    static final MethodHandle sqr8stride$mh = downcallHandle("sqr8s_stride", FunctionDescriptor.of(JAVA_INT));
+    @Override
+    public VectorSimilarityFunctions getVectorSimilarityFunctions() {
+        return INSTANCE;
+    }
 
-    static final MethodHandle dot8s$mh = downcallHandle("dot8s", FunctionDescriptor.of(JAVA_INT, ADDRESS, ADDRESS, JAVA_INT));
-    static final MethodHandle sqr8s$mh = downcallHandle("sqr8s", FunctionDescriptor.of(JAVA_INT, ADDRESS, ADDRESS, JAVA_INT));
+    private static final class JdkVectorSimilarityFunctions implements VectorSimilarityFunctions {
 
-    // Stride of the native implementation - consumes this number of bytes per loop invocation.
-    // There must be at least this number of bytes/elements available when going native
-    static final int DOT_STRIDE = 32;
-    static final int SQR_STRIDE = 16;
+        static final MethodHandle dot8stride$mh = downcallHandle("dot8s_stride", FunctionDescriptor.of(JAVA_INT));
```

Comment:
@ChrisHegarty GH made a bit of a mess with the diff, but essentially code from here to the end is the same, just indendeted (encapsulated in the private inner class)